src/hash.cr

```diff
@@ -1753,13 +1753,24 @@ class Hash(K, V)
   end
 
   # Compares with *other*. Returns `true` if all key-value pairs are the same.
-  def ==(other : Hash)
-    return false unless size == other.size
-    each do |key, value|
-      entry = other.find_entry(key)
-      return false unless entry && entry.value == value
-    end
-    true
+  def ==(other : Hash(K2, V2)) forall K2, V2
+    {% if K <= K2 %}
+      return false unless size == other.size
+      each do |key, value|
+        entry = other.find_entry(key.as(K))
+        return false unless entry && entry.value == value
+      end
+      true
+    {% elsif K2 < K %}
+      return false unless size == other.size
+      each do |key, value|
+        entry = other.find_entry(key.as(K2))
+        return false unless entry && entry.value == value
+      end
+      true
+    {% else %}
+      false
+    {% end %}
   end
```

Comment:
`Hash#==` would then fail if the argument is a union of `Hash` instances, so one level of indirection is needed here:

```crystal
class Hash(K, V)
  def ==(other : Hash)
    other.equals_impl(self)
  end

  protected def equals_impl(other : Hash(K2, V2)) forall K2, V2
    # ...
  end
end
```